crates/wasmtime/src/module/serialization.rs

```diff
@@ -0,0 +1,715 @@
+
+        self.check_version()?;
+        self.check_triple(isa)?;
+        self.check_isa_flags(isa)?;
+        self.check_strategy(compiler)?;
+        self.check_tunables(compiler)?;
+        self.check_features(compiler)?;
+
+        // Check the flags last as they are the least helpful in terms of diagnostic message
+        self.check_flags(isa)?;
+
+        let types = self
+            .tables
+            .into_iter()
+            .map(|t| Arc::new(t.unwrap_owned()))
+            .collect::<Vec<_>>();
+        let module = mk(engine, &types, self.data)?;
+
+        // Validate the module can be used with the current allocator
+        engine.allocator().validate(module.inner.module.module())?;
+
+        return Ok(module);
+
+        fn mk(
+            engine: &Engine,
+            types: &Vec<Arc<TypeTables>>,
+            data: SerializedModuleData<'_>,
+        ) -> Result<Module> {
+            let mut artifacts = CompiledModule::from_artifacts_list(
+                data.artifacts
+                    .into_iter()
+                    .map(|i| i.unwrap_owned())
+                    .collect(),
+                engine.compiler().isa(),
+                &*engine.config().profiler,
+            )?;
+            let inner = ModuleInner {
+                engine: engine.clone(),
+                types: types[data.type_tables].clone(),
+                module: artifacts.pop().unwrap(),
+                artifact_upvars: artifacts,
+                module_upvars: data
+                    .modules
+                    .into_iter()
+                    .map(|m| mk(engine, types, m))
+                    .collect::<Result<Vec<_>>>()?,
+            };
+
+            Ok(Module {
+                inner: Arc::new(inner),
+            })
+        }
+    }
+
+    fn check_version(&self) -> Result<()> {
+        if self.version != env!("CARGO_PKG_VERSION") {
+            bail!(
+                "Module was compiled with Wasmtime version '{}'",
+                self.version
+            );
+        }
+
+        Ok(())
+    }
+
+    fn check_triple(&self, isa: &dyn TargetIsa) -> Result<()> {
+        let triple = target_lexicon::Triple::from_str(&self.target).map_err(|e| anyhow!(e))?;
+
+        if triple.architecture != isa.triple().architecture {
+            bail!(
+                "Module was compiled for architecture '{}'",
+                triple.architecture
+            );
+        }
+
+        if triple.operating_system != isa.triple().operating_system {
+            bail!(
+                "Module was compiled for operating system '{}'",
+                triple.operating_system
+            );
+        }
+
+        Ok(())
+    }
+
+    fn check_flags(&self, isa: &dyn TargetIsa) -> Result<()> {
+        let host_level = isa.flags().opt_level().into();
+        if self.opt_level != host_level {
+            bail!("Module was compiled with optimization level '{:?}' but '{:?}' is expected for the host", self.opt_level, host_level);
+        }
```

Comment:
Do you think it's worth ignoring this? Out of the codegen options this is definitely one where it doesn't actually affect the output in a super meaningful way. An unoptimized `Engine` could easily take an optimized precompiled wasm module

Reply:
Agreed, there's no reason this can't be done.  The intention behind this was to keep the previous requirement of having the common flags match 1:1, but loosen the requirement behind the ISA flags as they are (currently) just CPU features and having the feature present on the host running the code (e.g. "I have AVX") while not targeted by AOT compilation (i.e. "I don't use AVX") should be allowed.

I can remove this check and allow the optimization level to differ (although I think it's a bit weird to say "I want to run this module as 'none' when it was already compiled optimized").